src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs

```diff
@@ -201,15 +189,34 @@ internal static bool TryUseHttps(this ListenOptions listenOptions)
     /// <returns>The <see cref="ListenOptions"/>.</returns>
     public static ListenOptions UseHttps(this ListenOptions listenOptions, HttpsConnectionAdapterOptions httpsOptions)
     {
-        var loggerFactory = listenOptions.KestrelServerOptions?.ApplicationServices.GetRequiredService<ILoggerFactory>() ?? NullLoggerFactory.Instance;
+        return listenOptions.UseHttps(() => httpsOptions);
+    }
 
+    /// <summary>
+    /// Configure Kestrel to use HTTPS. This does not use default certificates or other defaults specified via config or
+    /// <see cref="KestrelServerOptions.ConfigureHttpsDefaults(Action{HttpsConnectionAdapterOptions})"/>.
+    /// </summary>
+    /// <param name="listenOptions">The <see cref="ListenOptions"/> to configure.</param>
+    /// <param name="getHttpsOptions">A function returning options to configure HTTPS.</param>
+    /// <returns>The <see cref="ListenOptions"/>.</returns>
+    private static ListenOptions UseHttps(this ListenOptions listenOptions, Func<HttpsConnectionAdapterOptions> getHttpsOptions)
+    {
+        // Set this outside the lambda, which might not be invoked
         listenOptions.IsTls = true;
-        listenOptions.HttpsOptions = httpsOptions;
 
+        // NB: This lambda will only be invoked if either HTTP/1.* or HTTP/2 is being used
```

Comment:
What happens if there is an HTTP/3-only endpoint? Can it still pick up a certificate specified via UseHttps() now that listenOptions.HttpsOptions is only set in this non-HTTP/3 connection middleware?

Reply:
It couldn't when you wrote this, but it can now.